### ISSUE TITLE
Live Input Stream Interruptions and PTS reset

### ISSUE THREAD

**Issue by hariszukanovic** [google/shaka-packager#167]
When input stream is interrupted and as far as I see the pts numbers are reset, shaka packager starts issuing errors about segments being out of order. Packager does not crash or exit, but the packaged result is no longer playable after this.

Does anyone see any other valuable info from the below log?
Should the packager be able to survive such PTS reset on input?

[1025/193706:INFO:demuxer.cc(60)] Initialize Demuxer for file 'udp://238.100.100.6:6000'.
[1025/193711:INFO:demuxer.cc(176)] Demuxer::Run() on file 'udp://238.100.100.6:6000'.
[1025/193711:WARNING:mpd_builder.cc(233)] Video info does not contain frame duration required for calculating framerate. @frameRate is required for DASH IOP.
[1025/193711:WARNING:mpd_builder.cc(233)] Video info does not contain frame duration required for calculating framerate. @frameRate is required for DASH IOP.
[1025/193711:WARNING:mpd_builder.cc(233)] Video info does not contain frame duration required for calculating framerate. @frameRate is required for DASH IOP.
[1025/234843:WARNING:mpd_builder.cc(1265)] Found a gap of size 2160 > kRoundingErrorGrace (5). The new segment starts at 1716765600 but the previous segment ends at 1716763440.
[1025/235331:WARNING:mpd_builder.cc(1265)] Found a gap of size 2160 > kRoundingErrorGrace (5). The new segment starts at 1741981200 but the previous segment ends at 1741979040.
[1025/235540:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 126000 but the previous segment starts at 1741981200.
[1025/235540:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 129840 but the previous segment starts at 1754229840.
[1025/235540:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 129840 but the previous segment starts at 1754229840.
[1025/235540:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 129840 but the previous segment starts at 1754229840.
[1025/235546:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 984240 but the previous segment ends at 1026480.
[1026/000814:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 68520234 but the previous segment ends at 68520240.
[1026/002605:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 164871585 but the previous segment ends at 164871591.
[1026/003445:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 211696539 but the previous segment ends at 211696545.
[1026/005346:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 314351250 but the previous segment ends at 314351256.
[1026/010316:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 365678604 but the previous segment ends at 365678610.
[1026/012617:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 489944835 but the previous segment ends at 489944841.
[1026/013637:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 545774589 but the previous segment ends at 545774595.
[1026/015348:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 638524020 but the previous segment ends at 638524026.
[1026/020338:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 691652334 but the previous segment ends at 691652340.
[1026/022128:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 788003685 but the previous segment ends at 788003691.
[1026/023109:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 840231519 but the previous segment ends at 840231525.
[1026/024859:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 936582870 but the previous segment ends at 936582876.
[1026/025820:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 987009744 but the previous segment ends at 987009750.
[1026/032210:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 1115778375 but the previous segment ends at 1115778381.
[1026/033351:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 1178811969 but the previous segment ends at 1178811975.
[1026/035331:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 1285068600 but the previous segment ends at 1285068606.
[1026/040652:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 1357106994 but the previous segment ends at 1357107000.
[1026/050033:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 1647059831 but the previous segment ends at 1647061551.
[1026/055053:WARNING:mpd_builder.cc(1265)] Found a gap of size 2160 > kRoundingErrorGrace (5). The new segment starts at 1917205991 but the previous segment ends at 1917203831.
[1026/055618:WARNING:mpd_builder.cc(1265)] Found a gap of size 2160 > kRoundingErrorGrace (5). The new segment starts at 1943322071 but the previous segment ends at 1943319911.
[1026/055643:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 126000 but the previous segment starts at 1943322071.
[1026/055646:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 129840 but the previous segment starts at 1948629840.
[1026/055646:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 129840 but the previous segment starts at 1948629840.
[1026/055646:ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment. Adding segment with start_time == 129840 but the previous segment starts at 1948629840.
[1026/055650:WARNING:mpd_builder.cc(1276)] Segments should not be overlapping. The new segment starts at 715440 but the previous segment ends at 1026480.

**Comment by kqyang** [comment]
I don't have a good solution for this problem in packager right now. We don't want to drop frames or add new frames.

From the log, it seems that most of the time, the gaps or overlaps are quite small. Which player are you using? As far as I know, shaka player is able to survive small gaps/overlaps.

**Comment by hariszukanovic** [comment]
I patched it to make an exit on
ERROR:mpd_builder.cc(1255)] Segments should not be out of order segment
Also, I have it restart automatically because it runs as a service, and this seems to resolve the problem and the packager works fine after it is restarted.

Not really sure if this is a good solution though...
Nevertheless, perhaps shaka packager might have an array of what is considered unrecoverable errors and make a simple exit when those are encountered.

**Comment by kqyang** [comment]
Cool. Thanks for the update!

It does not look like a good solution at first glance, but as you said, we may have to make some difficult decisions on unrecoverable errors. We'll think about our options here.

Btw, do you also delete all previously generated segments on packager restart?

**Comment by kqyang** [comment]
Please ignore the question. I find the answer on #166. Thanks.

**Comment by hariszukanovic** [comment]
Of course, there is a definite interruption in service which is never good, but the interruption is upstream from shaka packager... so I am not sure how much can be done in Shaka packager to remedy it.
I haven't got deep enough knowledge of Shaka internals and/or how Dash is supposed to handle interruptions... to be able to propose nor imagine any kind of "good" improvement for this.
